Main.Test.hs

```diff
@@ -9,107 +9,200 @@ import Main
 
 -- | Unit tests the semantics of the conditional expression.
 --
---   >>> ifExprTest (B True) (I 5) (I 3)
+--   >>> ifExpr (B True) (I 5) (I 3) []
 --   I 5
 --
---   >>> ifExprTest (B False) (I 5) (I 3)
+--   >>> ifExpr (B False) (I 5) (I 3) []
 --   I 3
 --
---   >>> ifExprTest (B True) (S "abc") (I 123)
+--   >>> ifExpr (B True) (S "abc") (I 123) []
 --   S "abc"
 --
---   >>> ifExprTest (B False) (S "123") (B True)
+--   >>> ifExpr (B False) (S "123") (B True) []
 --   B True
 --
---   >>> ifExprTest (I 1) (S "123") (I 123)
+--   >>> ifExpr (I 1) (S "123") (I 123) []
 --   Error "Cannot branch expressions based on non-boolean condition"
 --
---   >>> ifExprTest (I 0) (S "123") (I 123)
+--   >>> ifExpr (I 0) (S "123") (I 123) []
```

Comment:
`ifExpr` now has additional `[]` because I have to pass env into it so I can use that env to evaluate `exprThen` and `exprElse` later down the line.

https://github.com/philectron/lil-lisp/blob/semantic/function_closure/Main.hs#L34-L37
```
ifExpr :: Expr -> Expr -> Expr -> Env Expr -> Expr
ifExpr (B cond) exprThen exprElse env = if cond then (expr exprThen env) else (expr exprElse env)
```

Due to performance and security reason, I don't evaluate those 2 expressions before the interpreter completes evaluating the `cond` expression